lib/puppet/provider/user/useradd.rb

```diff
@@ -147,19 +147,47 @@ def add_properties
     # validproperties is a list of properties in undefined order
     # sort them to have a predictable command line in tests
     Puppet::Type.type(:user).validproperties.sort.each do |property|
-      next if property == :ensure
-      next if property_manages_password_age?(property)
-      next if property == :groups and @resource.forcelocal?
-      next if property == :expiry and @resource.forcelocal?
-      # the value needs to be quoted, mostly because -c might
-      # have spaces in it
-      if value = @resource.should(property) and value != ""
-        cmd << flag(property) << munge(property, value)
-      end
+      value = get_value_for_property(property)
+      next if value.nil?
+      cmd << flag(property) << munge(property, value)
     end
     cmd
   end
 
+  def get_value_for_property(property)
+    return nil if property == :ensure
+    return nil if property_manages_password_age?(property)
+    return nil if property == :groups and @resource.forcelocal?
+    return nil if property == :expiry and @resource.forcelocal?
+    # the value needs to be quoted, mostly because -c might
+    # have spaces in it
+    value = @resource.should(property)
+    return nil if !value || value == ""
+
+    value
+  end
+
+  def has_sensitive_data?(property = nil)
+    properties = property ? [property] : Puppet::Type.type(:user).validproperties
+    properties.each do |prop|
+      value = get_value_for_property(prop)
+
+      if prop == :password
+        require 'byebug'; byebug
+      end
+
+      next if value.nil?
+
+      if @resource.property(prop).sensitive
+        require 'byebug'; byebug
+      end
+
+      return true if @resource.property(prop).sensitive
+    end
+
+    false
+  end
```

Comment:
Have to be careful with `return` from within a block (aka non-local return from a block), as it causes the enclosing `has_sensitive_data` method to return. In this case it's ok, because there's no code after `properties.each` and when the `has_sensitive_data` ends. But to avoid surprises it's best to use Ruby's `Enumerator#any?` to short-circuit evaluation once we find the first sensitive property:

```ruby
properties.any? do |prop|
  value = ...
  value ? @resource.property(prop).sensitive : false
end
```